pkg/cloudprovider/providers/openstack/openstack_routes.go

```diff
@@ -276,3 +277,41 @@ func (r *Routes) DeleteRoute(clusterName string, route *cloudprovider.Route) err
 	onFailure.Disarm()
 	return nil
 }
+
+func getPortByID(client *gophercloud.ServiceClient, portID string) (neutronports.Port, error) {
+	var port neutronports.Port
+	targetPort, err := neutronports.Get(client, portID).Extract()
+	if err != nil {
+		return port, err
+	}
+
+	if targetPort == nil {
+		return port, ErrNotFound
+	}
+
+	port = *targetPort
+	return port, nil
+}
+
+func getPortByIP(compute *gophercloud.ServiceClient, network *gophercloud.ServiceClient,
```

Comment:
This is now a rather deep function, that (eg) does openstack lookups several levels into the callstack.  From experience, this is going to lead to duplicate lookups for things like `getServerByName`, additional code for slightly different variations, etc.

Stepping back, we want to do some rather simple things here:
- Find the ports on `route.TargetNode`
- Find which one of those ports has the address `addr`

I think it's reasonable to expose the individual steps here rather than hide them within nested functions so we can reuse those building blocks efficiently elsewhere.  So build reusable pieces for the server lookups, and then change `getPortByIP` into a local "shallow" helper that only exists to avoid repeating ourselves across the Create+Delete cases above - something like:
```go
func getPortIDByIP(...) {
   // skipped error handling for clarity, but you get the idea
   srv := getServerByName(compute, targetNode)
   interfaces := getAttachedInterfacesByID(compute, srv.ID)
   var portID string
   for _, intf := range interfaces {
      for _, fixedIP : = range intf.FixedIPs {
         if fixedIP.IPAddress == ipAddress {
            portID = intf.PortID
            break
         }
      }
   }
   if portID == "" {
      return ErrNotFound
   }
   return portID
}
```